### PR TITLE
Support cudnn in load_lua

### PR DESCRIPTION
**Pull request by szagoruyko**
A couple of dirty hacks to support cudnn. Main problems:
* there's no torch.cuda.FloatStorage.from_buffer so reading fails on CudaStorages https://github.com/pytorch/pytorch/blob/master/torch/utils/serialization/read_lua_file.py#L163
* there's no torch.CudaTensor backend, I added a fallback to torch.FloatTensor

### CODE REVIEW

**Review by apaszke — CHANGES_REQUESTED** [review]

File: torch/utils/serialization/read_lua_file.py [hunk 1]
```diff
@@ -160,7 +160,11 @@ def read_storage(reader, version):
         # source:
         # https://github.com/torch/torch7/blob/master/generic/Storage.c#L244
         size = reader.read_long() * element_size
-        return python_class.from_buffer(reader.f.read(size), 'native')
+        # because there's no torch.cuda.FloatStorage.from_buffer
+        if python_class == torch.cuda.FloatStorage:
+            return torch.FloatStorage.from_buffer(reader.f.read(size), 'native').cuda()
+        else:
+            return python_class.from_buffer(reader.f.read(size), 'native')
     return read_storage
```
Comment (Wakeupbuddy) [inline torch/utils/serialization/read_lua_file.py]: I think we need to call some decorators for cudnn modules as well, like:

`ensure_attr('cudnn.SpatialConvolution', 'bias', 'finput', 'fgradInput', 'gradWeight', 'gradBias')
ensure_attr('cudnn.SpatialBatchNormalization', 'weight', 'bias', 'gradWeight', 'gradBias')
ensure_attr('cudnn.BatchNormalization', 'weight', 'bias', 'gradWeight', 'gradBias')`

**Comment by soumith** [comment]
the PR needs to be reopened, as it now comes from a private branch

File: torch/utils/serialization/read_lua_file.py [hunk 2]
```diff
@@ -245,6 +250,19 @@ def read_nn_class(reader, version):
...
+        'SpatialAveragePooling': nn.SpatialAveragePooling,
+        'SpatialMaxPooling': nn.SpatialMaxPooling,
+        'SpatialBatchNormalization': nn.SpatialBatchNormalization,
+        'BatchNormalizaion': nn.BatchNormalization,
+        }
+
+reader_registry.update({('cudnn.' + name): nn_reader(module)
+    for name, module in cudnn.items()
+    if name[0] != '_' and name[0].upper() == name[0]})
+
```
[the remark this answers is not in the document]
Comment (WayneZww) [reply torch/utils/serialization/read_lua_file.py]: But where were these code gone? I find the currently code dose not contain these lines and I still got error when loading model contains cudnn.SpatialConvolution
